src/main/java/org/opensearch/geospatial/index/query/xyshape/XYShapeQueryProcessor.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.index.query.xyshape;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Objects;
+
+import org.apache.lucene.document.ShapeField;
+import org.apache.lucene.document.XYShape;
+import org.apache.lucene.geo.XYGeometry;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.opensearch.common.geo.ShapeRelation;
+import org.opensearch.geometry.Geometry;
+import org.opensearch.geometry.GeometryVisitor;
+import org.opensearch.geospatial.index.mapper.xyshape.XYShapeFieldMapper;
+import org.opensearch.index.mapper.MappedFieldType;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.index.query.QueryShardException;
+
+/**
+ * Query Processor to convert given Geometry into Lucene query
+ */
+public class XYShapeQueryProcessor {
+
+    /**
+     * Creates a {@link Query} that matches all indexed shapes to the provided {@link Geometry}  based on {@link ShapeRelation}
+     * @param geometry OpenSearch {@link Geometry} as an input
+     * @param fieldName Lucene field of type {@link XYShape}
+     * @param relation Relation to be used get all shapes from giben Geometry
+     * @param context QueryShardContext instance
+     * @return {@link Query} instance from XYShape.newGeometryQuery
+     */
+    public Query shapeQuery(
+        Geometry geometry,
+        String fieldName,
+        ShapeRelation relation,
+        GeometryVisitor<List<XYGeometry>, RuntimeException> visitor,
+        QueryShardContext context
+    ) {
+        if (geometry == null || geometry.isEmpty()) {
+            return new MatchNoDocsQuery();
+        }
+        Objects.requireNonNull(geometry, "Field name cannot be null");
+        Objects.requireNonNull(geometry, "ShapeRelation cannot be null");
+        Objects.requireNonNull(geometry, "QueryShardContext cannot be null");
+        validateIsXYShapeFieldType(fieldName, context);
+        return getQueryFromGeometry(geometry, fieldName, relation.getLuceneRelation(), visitor);
+    }
+
+    private Query getQueryFromGeometry(
+        Geometry geometry,
+        String fieldName,
+        ShapeField.QueryRelation queryRelation,
+        GeometryVisitor<List<XYGeometry>, RuntimeException> visitor
+    ) {
+        final List<XYGeometry> collections = geometry.visit(visitor);
+        return XYShape.newGeometryQuery(fieldName, queryRelation, collections.toArray(new XYGeometry[0]));
```

Comment:
Should we check for collections isEmpty?

Reply:
Good point. Added condition and included unit tests